scubagoggles/auth.py

```diff
@@ -4,54 +4,167 @@
 from google_auth_oauthlib.flow import InstalledAppFlow
 
-# If modifying these scopes, delete the file token.json.
-SCOPES = ['https://www.googleapis.com/auth/admin.reports.audit.readonly',
-    "https://www.googleapis.com/auth/admin.directory.domain.readonly",
-    "https://www.googleapis.com/auth/admin.directory.orgunit.readonly",
-    "https://www.googleapis.com/auth/admin.directory.user.readonly",
-    "https://www.googleapis.com/auth/admin.directory.group.readonly",
-    "https://www.googleapis.com/auth/apps.groups.settings"]
+# The class is worth it just for the encapsulation.  It allows the potential
+# of credential refresh multiple times, which may be beneficial during a
+# prolonged debugging session.
+# pylint: disable=too-few-public-methods
 
 
-def gws_auth(cred_path:str, subject_email: str = None):
-    """
-    Generates an Oauth token for accessing Google's APIs
+class GwsAuth:
 
-    :param cred_path: directory containing the credentials file
-    :param subject_email: if set, assumes credentials are for service account and uses
-        this email as the subject
+    """Generates an Oauth token for accessing Google's APIs
     """
-    cred_dir = Path(cred_path).parent
-    creds = None
-
-    if subject_email is not None:
-        creds = SvcCredentials.from_service_account_file(cred_path, scopes=SCOPES,
-                                                         subject=subject_email)
-        if not creds.valid:
-            creds.refresh(Request())
-        return creds
-
-    oauth_token = (cred_dir / 'token.json').resolve()
-    if os.path.exists(oauth_token):
-        creds = Credentials.from_authorized_user_file(oauth_token, SCOPES)
-    # If there are no (valid) credentials available, let the user log in.
-    if not creds or not creds.valid:
-        if creds and creds.expired and creds.refresh_token:
-            creds.refresh(Request())
-        else:
-            flow = InstalledAppFlow.from_client_secrets_file(
-                cred_path, SCOPES)
-            # the prompt parameter forces the user to consent to the token
-            creds = flow.run_local_server(port=8080, prompt='consent')
-        # Save the credentials for the next run
-        with open(oauth_token, mode='w', encoding='UTF-8') as token:
-            token.write(creds.to_json())
-    return creds
+
+    _base_auth_url = 'https://www.googleapis.com/auth'
+
+    _scopes = (f'{_base_auth_url}/admin.reports.audit.readonly',
+               f'{_base_auth_url}/admin.directory.domain.readonly',
+               f'{_base_auth_url}/admin.directory.orgunit.readonly',
+               f'{_base_auth_url}/admin.directory.user.readonly',
+               f'{_base_auth_url}/admin.directory.group.readonly',
+               f'{_base_auth_url}/apps.groups.settings',
+               f'{_base_auth_url}/cloud-identity.policies.readonly')
+
+    def __init__(self, credentials_path: Path, svc_account_email: str = None):
```

Comment:
Fix service account authentication:
- The `GwsAuth` determines if the creds file is for a service account by checking if the `svc_account_email` parameter is not `None`
- The `GwsAuth` has that parameter default to `None`
- The provider instantiates the auth class like so: `GwsAuth(credentials_file)`, meaning `svc_account_email` will always be `None`, even if the user provides one.
- This causes the auth class to improperly handle the service account credentials, resulting in `ValueError: Client secrets must be for a web or installed app.`

Reply:
done - this should work - I don't have access to a service account to test.

Reply:
I can confirm, service account authentication is working now, thanks for the updates.